### ISSUE TITLE
Gradle versions above 4.7 cannot resolve kotlinx-coroutines-core with GRADLE_METADATA feature enabled

### ISSUE THREAD

**Issue by serebit** [Kotlin/kotlinx.coroutines#564]
When attempting to resolve `kotlinx-coroutines-core`, Gradle versions above 4.7 show the following message:
```
> Could not resolve all files for configuration ':source:compileClasspath'.
   > Could not resolve org.jetbrains.kotlinx:kotlinx-coroutines-core:0.26.0.
     Required by:
         project :source
      > Could not resolve org.jetbrains.kotlinx:kotlinx-coroutines-core:0.26.0.
         > Could not parse module metadata https://jcenter.bintray.com/org/jetbrains/kotlinx/kotlinx-coroutines-core/0.26.0/kotlinx-coroutines-core-0.26.0.module
            > Unsupported format version '0.3' specified in module metadata. This version of Gradle supports format version 0.4 only.
```

**Comment by JakeWharton** [comment]
I don't have this problem with 4.10 or 4.10.1.

**Comment by serebit** [comment]
> I don't have this problem with 4.10 or 4.10.1.

If the dependency is already cached in `~/.gradle`, this doesn't happen. Delete the user gradle cache and see if it happens then.

**Comment by JakeWharton** [comment]
I used either 4.9 or 4.10 to get it in the first place though. In any case, trying now!

**Comment by JakeWharton** [comment]
No problems. I also banish jcenter to be very last in my repository list so i'm likely pulling from maven central. Perhaps try adding mavenCentral first?

**Comment by serebit** [comment]
Tried adding mavenCentral first, to no effect. Let me try something else... This persists across two PCs I have, so I know it isn't just the PC I'm using.

**Comment by serebit** [comment]
Try cloning https://gitlab.com/serebit/diskord and attempt to `gradlew assemble`. If you don't see the error I posted above, there's something weird going on.

**Comment by serebit** [comment]
Actually, I'll reopen this, seeing as it is still an issue with the published library files.

**Comment by elizarov** [comment]
Let's keep this open. Gradle metadata is an experimental Gradle feature and its format changes between Gradle version. Since we publish our libraries with Gradle 4.7 and our builds have Gradle metadata feature enable, you cannot use it from any other version of Gradle if you also happen to have Gradle metadata enabled. We'll figure out how to publish our libraries _without_ Gradle metadata.

**Comment by serebit** [comment]
Sounds like a good course of action. But now I'm wondering why you don't just update Gradle to a newer version that supports metadata version 4 to tide us over while you figure out how to disable metadata.

**Comment by JakeWharton** [comment]
The Android Gradle plugin 3.3 requires Gradle 4.10 or higher which makes this especially problematic.

**Comment by elizarov** [comment]
The problem is worse than that. We'll have to pick a version and then we'll have to you this version everywhere, and then, when a new Gradle version, we'll have to rebuild everything. Long story short, we should not be using Gradle metadata until it is stable.

**Comment by budioktaviyan** [comment]
So, the best choice for now is stick to use the gradle 4.6 instead ?
p.s: I'm using android with gradle plugin `3.2.1` and coroutines version `0.30.2`, the problem still exist

**Comment by budioktaviyan** [comment]
[Update]  I'm using gradle `4.7` and no issue. also the `GRADLE_METADATA` are enabled

**Comment by serebit** [comment]
Update Gradle to any version after 4.7, clear your .gradle folders, and try again. It’ll fail.

**Comment by budioktaviyan** [comment]
`Before`
<img width="742" alt="screen shot 2018-10-30 at 05 30 48" src="https://user-images.githubusercontent.com/2031493/47684343-26e09600-dc05-11e8-86ac-17713f5db132.png">

`After`
<img width="948" alt="screen shot 2018-10-30 at 05 31 43" src="https://user-images.githubusercontent.com/2031493/47684357-33fd8500-dc05-11e8-8eb5-e1b5f3ef7e8f.png">

It was no issue from my side

**Comment by serebit** [comment]
Not just clean build. You have to *manually* delete your gradle caches at $HOME/.gradle and .gradle in your project directory.

**Comment by budioktaviyan** [comment]
Yeah, I also clear all my cache in `$GRADLE_HOME` before. There was no issue on my side

**Comment by serebit** [comment]
And what gradle version are you using?

**Comment by budioktaviyan** [comment]
Previously I'm using `4.10.2` and have the same problem like you, but after downgrading wrapper to `4.7` the problem was gone. Here is my project, maybe you can check it later (sorry, it still not a complete project since I'm just started build a sample) :)

**Comment by serebit** [comment]
Yeah, I’m aware that the problem doesn’t exist with 4.7. Downgrading to 4.7 fixed my issue, but only while I still had the dependency in the cache. If I switched back to 4.10.2 and cleared cache, it would happen again. That’s why the issue title mentions that the version has to be newer than 4.7.

**Comment by luca992** [comment]
Are there any updates on this now that Android Studio 3.3 is in it's release candidate stage, and is close to being released? Also, does Gradle 5.0 help this issue at all since GRADLE_METADATA is enabled by default?

**Comment by jonl-percsolutions-com** [comment]
Android 3.3 now released today and upgrading and following prompt recommendations will lead you down a rabbit hole.

**Comment by 4face-studi0** [comment]
I would also add that **SqlDelight 1.0** requires Gradle Metadata :(

**Comment by JakeWharton** [comment]
That's not quite true. You can specify the common and each platform dependency manually with it disabled.

**Comment by 4face-studi0** [comment]
So I've been tricked :D https://github.com/square/sqldelight/issues/1149 . I'll try to figure out a solution

**Comment by luca992** [comment]
@JakeWharton Wow, I never realized enabling Gradle metadata was actually optional if you manually specify the dependencies... This could have saved me so much fustration these past couple montha during this transitional period. Thanks for the tip

**Comment by dennisdrew** [comment]
@luca992 @JakeWharton what dependencies are we talking about manually specifying? Other than the platform-specific ones (i.e. `coroutines-ios`) being where they belong, and the common ones being where they belong, what needs to change if I disable metadata?

**Comment by luca992** [comment]
@dennisdrew Basically for example take sqldelight runtime:

with metadata enabled: `enableFeaturePreview('GRADLE_METADATA')`

    kotlin {
        sourceSets {
            commonMain {
                kotlin.srcDirs +="build/sqldelight"
                dependencies {
                    implementation "com.squareup.sqldelight:runtime:$sqldelight_version"
                    }
            }
            iosMain
        }
    
        targets {
            fromPreset(presets.iosX64, 'ios')
        }
    }


without metadata enabled becomes:

    kotlin {
        sourceSets {
            commonMain {
                kotlin.srcDirs +="build/sqldelight"
                dependencies {
                    implementation "com.squareup.sqldelight:runtime-metadata:$sqldelight_version"
                    }
            }
            iosMain {
                dependsOn commonMain
                dependencies {
                    implementation "com.squareup.sqldelight:runtime-iosarm64:$sqldelight_version"
                }
            }
        }
    
        targets {
            fromPreset(presets.iosArm64, 'ios')
        }
    }

**Comment by dennisdrew** [comment]
thanks @luca992. that makes sense and is helpful! Do you know if something similar has to be done in order to get coroutines to compile? I disabled metadata and now coroutines doesn't compile when building for iOS

**Comment by dennisdrew** [comment]
@luca992 tried :( but still not working. I don't want to flood this thread w/ this stuff--if you're on kotlinlang.slack and still feel like helping let me know, or shoot me an e-mail

**Comment by qwwdfsad** [comment]
Since version 1.1.1 kotlinx.corouintes are published with Gradle 4.10 and new metadata (0.4)
